pkg/dockerregistry/server/blobdescriptorservice.go

```diff
@@ -25,6 +38,137 @@ type blobDescriptorService struct {
 	distribution.BlobDescriptorService
 }
 
+// Stat returns a a blob descriptor if the given blob is either linked in repository or is referenced in
+// corresponding image stream. This method is invoked from inside of upstream's linkedBlobStore. It expects
+// a proper repository object to be set on given context by upper openshift middleware wrappers.
 func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	// if there is a repo layer link, return its desc
+	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
```

Comment:
So the *only* place in memory cache was being used was here?  No one else
is expecting to make blob store calls that are inexpensive (because they're
guarded by the cache) but will only find out that they are very expensive?

On Sat, Jul 16, 2016 at 11:09 AM, Michal Minar <notifications@github.com>
wrote:

> In pkg/dockerregistry/server/blobdescriptorservice.go
> <https://github.com/openshift/origin/pull/9819#discussion_r71067249>:
>
> >  func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
> > -	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
> > +	// if there is a repo layer link, return its desc
> > +	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
>
> We may need to reorder it rather than turning it off - my worry is that
> lower code depends on it.
>
> Not really. The order of wrappers and stores is:
>
> errorBlobStore { pullthroughBlobStore { quotaRestrictedBlobStore { linkedBlobStore { cachedBlobStatter {
>    cache: inMemoryBlobDescriptorCacheProvider,
>    backend: blobDescriptorService { upstreamBlobDescriptorService },
> }}}}}
>
> The inMemoryCache will be updated with any successful stat we return with
> our backend and the result will be kept forever. Reordering (moving our
> blobDescriptorService one level up) is not possible without patching
> upstream again. I'd rather use cachedLayers cache as a replacement for
> the inMemoryCache because in our case all the layer links and etcd data are
> ephemeral with proper periodic pruning.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9819/files/1bb6fba1b13913a5636eff319ceead8d79b9f50c#r71067249>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pxX13-171aNMXwXH223oFJ1Rf7SNks5qWPQ5gaJpZM4JLnQb>
> .
>